drivers/radios Apply register to rssi dBm conversion to all radios

bergzand opened the pull request:
Implementation idea of #6887. Should be easy to convert to absolute positive values if preferred.

Hyungsin commented:
@bergzand, is it possible for you to change the implementation according to @gebart'comment in #6887 ?
Using int16_t for rssi seems to be good. Any concern about this change?

bergzand commented:
@Hyungsin I've got no problem with that, but I'd like some kind of GO from somebody higher up in the hierarchy before I'm going to change and check most of the radio drivers

Hyungsin commented:
@thomaseichinger, any feedback?

thomaseichinger commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -166,15 +166,17 @@ static int _recv(netdev_t *netdev, void *buf, size_t len, void *info)
-        radio_info->rssi = at86rf2xx_reg_read(dev, AT86RF2XX_REG__PHY_ED_LEVEL);
+        rssi = at86rf2xx_reg_read(dev, AT86RF2XX_REG__PHY_ED_LEVEL);
 #endif
+        radio_info->rssi = -RSSI_BASE_VAL - rssi;
```
@bergzand Please help me out, what is the reason for changing the algebraic sign of this calculation? Especially as you are changing `.rssi` to `int8_t` ...

bergzand replied:
I forgot to change `.rssi` back to the original `uint8_t`. The sign here is flipped to have the value from the datasheet directly as a define. The sign is flipped to correct the algabraic formula from sec 8.5.3 from the datasheet of the at86rf233 to get a positive value.

thomaseichinger commented on drivers/cc2420/include/cc2420_registers.h:
```diff
@@ -234,6 +234,13 @@ enum {
+ * @{
+ */
+#define CC2420_CRCCOR_CRC_MASK      (0x80)
+#define CC2420_CRCCOR_COR_MASK     (0x7F)
```
missing `/** @} */`

thomaseichinger commented on drivers/cc2420/cc2420.c:
```diff
@@ -200,14 +201,18 @@ int cc2420_rx(cc2420_t *dev, uint8_t *buf, size_t max_len, void *info)
         /* fetch and check if CRC_OK bit (MSB) is set */
         cc2420_fifo_read(dev, &crc_corr, 1);
-        if (!(crc_corr & 0x80)) {
+        if (!(crc_corr & CC2420_CRCCOR_CRC_MASK)) {
```
Think this and related changes should go in a separate commit.

bergzand replied:
Good point.

bergzand commented:
@thomaseichinger Thank you for your feedback. I initially build the implementation here to have rssi as a `int8_t`, but changed it to `uint8_t` after some discussion in #6887. I think I only changed the implementation to report rssi as the absolute value, but forgot to change the variable back to `uint8_t`.

bergzand commented:
The main discussion point now is whether to use `uint8_t` to express RSSI as the absolute value or as proposed by @gebart use `int16_t` for RSSI.

thomaseichinger commented:
@bergzand Thank you for shedding some light on it. Personally I think @gebart's suggestion using `int16_t` provides us with more flexibility.

bergzand commented:
Great, then I make sure to adjust and clean up this PR. Is it okay to keep it one big PR for all radio drivers or do you prefer separate PR's for each radio?

thomaseichinger commented:
I think one PR is fine but maybe split changes in one commit per device.

bergzand commented:
@thomaseichinger I rebased it a bit, it is now a set of commits, one for the `.rssi` and one per radio. `.rssi` is now `int16_t` and I corrected all issues from your review.

thomaseichinger commented:
Awesome! Finally, could you add a mention in the header comment that rssi has to be in dBm?

bergzand commented:
Of course.
Added "in dBm" to the comments of the two structs.

Hyungsin commented:
Cool! Looking forward to its merge :)

bergzand commented:
Rebased to resolve merge conflict

bergzand commented:
Polite ping :)

jia200x commented:
Looking forward to this feature 👍

Hyungsin commented:
@thomaseichinger May be you can confirm this PR :)

smlng commented on drivers/cc2420/cc2420.c:
```diff
@@ -197,8 +197,8 @@ int cc2420_rx(cc2420_t *dev, uint8_t *buf, size_t max_len, void *info)
 
-        uint8_t rssi;
-        cc2420_fifo_read(dev, &rssi, 1);
+        int8_t rssi;
```
why change this to signed, and then cast it to unsigned below? Seems a bit odd to me.

bergzand replied:
Probably a leftover from the first implementation that used `uint8_t`s. Will fix.

bergzand replied:
Ignore my previous comment. @smlng Assuming you're referring to the cast in the `cc2420_fifo_read()` function, that function requires a `uint8_t*` to write the result of the read to. The value from the cc2420 register should be interpreted as "signed 2’s complement".

smlng replied:
ah okay, understood

smlng commented on drivers/include/net/netdev.h:
```diff
@@ -245,7 +245,7 @@ typedef enum {
  */
 struct netdev_radio_rx_info {
-    uint8_t rssi;       /**< RSSI of a received packet */
+    int16_t rssi;       /**< RSSI of a received packet in dBm */
```
is it really necessary to increase it that much, as far as I understand value can be negative, but typically would range from -100 to +100 or something like that. While `uint8_t` is clearly wrong for all possible RSSI values, IMHO `int8_t` would suffice.

bergzand replied:
This was discussed in #6887. The arguments there against `int8_t` is that LoRa(-like) receivers can work with values below -128dBm.

smlng replied:
understood, then it makes sense to allow for a larger value range here.

smlng commented on drivers/mrf24j40/mrf24j40_getset.c:
```diff
@@ -74,38 +74,38 @@ static const uint8_t dbm_to_tx_pow[] = { 0x00, 0x10, 0x18, 0x20, 0x28, 0x30, 0x3
-                                     41, 41, 41, 41, 41, 40, 40, 40, \
-                                     40, 40, 39, 39, 39, 39, 39, 38, \
-                                     38, 38, 38, 37, 37, 37, 36, 30 };
+static const int16_t dBm_value[] = { -90, -89, -88, -88, -87, -87, -87, -87, \
```
looking at the values I don't see why we should double the amount of memory required here

smlng replied:
plus: why change (or correct) first and last value by 5?

smlng replied:
even when considering the increase value range for the info struct, its IMHO no necessary to do so here, specifically as the values fit into `int8_t` and `int16_t` would double the required memory <- bad.

bergzand replied:
> looking at the values I don't see why we should double the amount of memory required here

Ack, will fix.

> plus: why change (or correct) first and last value by 5?

That was a small correction from the datasheet values, from the worst (-100dBm) truncated value to the first truncated (-90dBm) value. I'll can put it in a separate PR if you want.

smlng replied:
no need to put it in a separate PR, was just curious.

bergzand replied:
The whole range from -100dBm to -90dBm is returned as a value of 0, the same with -35dBm to -25dBm. To me it looked weird to have a discontinuity in the values by using the worst/optimal case.

smlng replied:
yeah, just checked the data sheet, I guess the intention here was to use the mean of those same value upper and lower bounds.

smlng replied:
sadly there is no simple function to optimise this lookup table a bit or at best completely away.

smlng commented on sys/include/net/gnrc/netif/hdr.h:
```diff
@@ -85,7 +85,7 @@ typedef struct {
     kernel_pid_t if_pid;        /**< PID of network interface */
     uint8_t flags;              /**< flags as defined above */
-    uint8_t rssi;               /**< rssi of received packet (optional) */
+    int16_t rssi;               /**< rssi of received packet in dBm (optional) */
```
as above IMHO `int8_t` should do

smlng replied:
change is fine, as discussed above

smlng commented:
I would suggest to use the smallest fitting data type in the low-lever driver, and use casts where necessary to put data into common info struct.

bergzand commented:
@smlng Thanks for the review.

> I would suggest to use the smallest fitting data type in the low-lever driver, and use casts where necessary to put data into common info struct.

Yes, that makes sense, I will check the radio code and fix where necessary.

bergzand commented:
@smlng addressed your comments.

Oh, and I just noticed that by touching the gnrc_netif header, this PR is restricted by the embargo :'(

gvz commented:
Ok, have not seen it, the lqi was more important to me, so I searched for that.

bergzand commented:
@gvz If you have time, could you please test this PR with the cc2420? I don't have access to that radio, so my code is a best guess on the information from the datasheet.

gvz commented:
@bergzand Just tried on the cc2420 and the rssi is 65470, I guess something went wrong. I will try to have a look.

bergzand commented:
@gvz That looks like an interpretation error, `int16_t` as `uint16_t`. 65470 is -66 when interpreted as signed, and that looks like a valid number for rssi.

gvz commented:
Ok I found it: its in: gnrc_netif_hdr_print.c the fix is:
- printf("rssi: %u  ", (unsigned) hdr->rssi);
+ printf("rssi: %d  ", (signed) hdr->rssi);

bergzand commented:
Got it, thanks, will put a fixup asap here.

bergzand commented:
@gvz It should print negative numbers now :)

smlng commented:
please squash

bergzand commented:
Squashed
